### PR TITLE
Implement PBR anisotropy per `KHR_materials_anisotropy`.

### PR DESCRIPTION
**Pull request by pcwalton**
This commit implements support for physically-based anisotropy in Bevy's `StandardMaterial`, following the specification for the [`KHR_materials_anisotropy`] glTF extension.

[*Anisotropy*] (not to be confused with [anisotropic filtering]) is a PBR feature that allows roughness to vary along the tangent and bitangent directions of a mesh. In effect, this causes the specular light to stretch out into lines instead of a round lobe. This is useful for modeling brushed metal, hair, and similar surfaces. Support for anisotropy is a common feature in major game and graphics engines; Unity, Unreal, Godot, three.js, and Blender all support it to varying degrees.

Two new parameters have been added to `StandardMaterial`: `anisotropy_strength` and `anisotropy_rotation`. Anisotropy strength, which ranges from 0 to 1, represents how much the roughness differs between the tangent and the bitangent of the mesh. In effect, it controls how stretched the specular highlight is. Anisotropy rotation allows the roughness direction to differ from the tangent of the model.

In addition to these two fixed parameters, an *anisotropy texture* can be supplied. Such a texture should be a 3-channel RGB texture, where the red and green values specify a direction vector using the same conventions as a normal map ([0, 1] color values map to [-1, 1] vector values), and the the blue value represents the strength. This matches the format that the [`KHR_materials_anisotropy`] specification requires. Such textures should be loaded as linear and not sRGB. Note that this texture does consume one additional texture binding in the standard material shader.

The glTF loader has been updated to properly parse the `KHR_materials_anisotropy` extension.

A new example, `anisotropy`, has been added. This example loads and displays the barn lamp example from the [`glTF-Sample-Assets`] repository. Note that the textures were rather large, so I shrunk them down and converted them to a mixture of JPEG and KTX2 format, in the interests of saving space in the Bevy repository.

[*Anisotropy*]: https://google.github.io/filament/Filament.md.html#materialsystem/anisotropicmodel

[anisotropic filtering]: https://en.wikipedia.org/wiki/Anisotropic_filtering

[`KHR_materials_anisotropy`]: https://github.com/KhronosGroup/glTF/blob/main/extensions/2.0/Khronos/KHR_materials_anisotropy/README.md

[`glTF-Sample-Assets`]: https://github.com/KhronosGroup/glTF-Sample-Assets/

### Added

* Physically-based anisotropy is now available for materials, which enhances the look of surfaces such as brushed metal or hair. glTF scenes can use the new feature with the `KHR_materials_anisotropy` extension.

## Screenshots

With anisotropy:
Screenshot 2024-05-20 233414

Without anisotropy:
Screenshot 2024-05-20 233420

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -3059,6 +3059,18 @@ description = "Demonstrates volumetric fog and lighting"
 category = "3D Rendering"
 wasm = true
 
+[[example]]
+name = "anisotropy"
+path = "examples/3d/anisotropy.rs"
+doc-scrape-examples = true
+required-features = ["jpeg"]
```
Comment (torsteingrindvik) [inline Cargo.toml:3066]: Should the `ktx2` feature be enabled too?

https://github.com/bevyengine/bevy/blob/main/Cargo.toml#L208

File: assets/models/AnisotropyBarnLamp/AnisotropyBarnLamp_basecolor.jpeg
Comment (torsteingrindvik) [inline assets/models/AnisotropyBarnLamp/AnisotropyBarnLamp_basecolor.jpeg:1]: Any particular reason to keep this as jpeg?

You likely know but: `basisu -ktx2 -uastc -mipmap foo.jpeg` has worked nicely for me to get mipmaps.

File: crates/bevy_pbr/src/render/pbr_fragment.wgsl [hunk 2]
```diff
@@ -418,6 +418,43 @@ fn pbr_input_from_standard_material(
...
+                uv_b,
+#else   // STANDARD_MATERIAL_ANISOTROPY_UV_B
+                uv,
+#endif  // STANDARD_MATERIAL_ANISOTROPY_UV_B
+                bias,
+            ).rgb;
+
+            let anisotropy_direction_from_texture = normalize(anisotropy_texel.rg * 2.0 - 1.0);
+            anisotropy_direction =
+                mat2x2(anisotropy_direction.xy, anisotropy_direction.yx * vec2(-1.0, 1.0)) *
```
Comment (torsteingrindvik) [inline crates/bevy_pbr/src/render/pbr_fragment.wgsl:443]: This looks like it's doing a common 2D rotation but I'll admit I can't remember it's not obvious to me that it's correct, could you add a comment?

Alternatively, things like `let TBN = fns::calculate_tbn_mikktspace(mesh.world_normal, mesh.world_tangent);` inspire confidence so perhaps a helper fn could be introduced.

File: Cargo.toml (hunk 1 above)
Comment (pcwalton) [reply Cargo.toml:3066]: It doesn't seem to require that for me. I think it's on by default?

File: assets/models/AnisotropyBarnLamp/AnisotropyBarnLamp_basecolor.jpeg
Comment (pcwalton) [reply assets/models/AnisotropyBarnLamp/AnisotropyBarnLamp_basecolor.jpeg:1]: I think JPEG results in the smallest possible file size, and there's been a lot of discussion about keeping example files small in the repo. Mipmaps would increase size, which I didn't think was worth it as the textures look pretty good and the performance is fine.

File: crates/bevy_gltf/src/loader.rs [hunk 3]
```diff
@@ -1868,10 +1881,52 @@ impl ClearcoatExtension {
...
+            anisotropy_channel: anisotropy_channel.unwrap_or_default(),
+            anisotropy_texture,
+        })
+    }
+}
+
 /// Returns the index (within the `textures` array) of the texture with the
 /// given field name in the data for the material extension with the given name,
 /// if there is one.
-#[cfg(feature = "pbr_multi_layer_material_textures")]
```
Comment (atlv24) [inline crates/bevy_gltf/src/loader.rs:1874]: why was this removed?

File: crates/bevy_pbr/src/pbr_material.rs [hunk 4]
```diff
@@ -467,11 +467,79 @@ pub struct StandardMaterial {
...
+    /// [`KHR_materials_anisotropy` specification]:
+    /// https://github.com/KhronosGroup/glTF/blob/main/extensions/2.0/Khronos/KHR_materials_anisotropy/README.md
+    pub anisotropy_strength: f32,
+
+    /// The direction of increased roughness, in radians relative to the mesh
+    /// tangent.
+    ///
+    /// This parameter causes the roughness to vary according to the
+    /// [`StandardMaterial::anisotropy_strength`]. The rotation is applied in
+    /// tangent-bitangent space; thus, mesh tangents mush be present for this
```
Comment (atlv24) [inline crates/bevy_pbr/src/pbr_material.rs:499]:
```suggestion
    /// tangent-bitangent space; thus, mesh tangents must be present for this
```

**Review by atlv24 — APPROVED** [review]
love the specular normal mixing trick, and the example is very good too. sweet beans

File: crates/bevy_gltf/src/loader.rs (hunk 3 above)
Comment (pcwalton) [reply crates/bevy_gltf/src/loader.rs:1874]: Anisotropy needs to use this function. Unlike clearcoat (the previous user of this function), anisotropy's textures aren't gated behind the `pbr_multi_layer_material_textures` feature, as it's not technically a separate layer.

File: Cargo.toml (hunk 1 above)
Comment (torsteingrindvik) [reply Cargo.toml:3090]: Just checked, seems `tonemapping_luts` implies `ktx2` and the former is enabled by default. Resolving this

**Comment by IceSentry** [comment]
The `volumetric_fog` example crashes with this PR. 

```
wgpu error: Validation Error

Caused by:
    In a RenderPass
      note: encoder = `<CommandBuffer-(4, 75, Vulkan)>`
    In a set_bind_group command
      note: bind group = `mesh_view_bind_group`
    Bind group 0 expects 5 dynamic offsets. However 4 dynamic offsets were provided.
```

File: crates/bevy_pbr/src/render/pbr_lighting.wgsl [hunk 5]
```diff
@@ -133,6 +145,30 @@ fn D_GGX(roughness: f32, NdotH: f32, h: vec3<f32>) -> f32 {
...
+// * `T` = 𝐭 = the tangent direction = the direction of increased roughness.
+//
+// * `B` = 𝐛 = the bitangent direction = the direction of decreased roughness.
+//
+// * `at` = αₜ = the alpha-roughness in the tangent direction.
+//
+// * `ab` = α_b = the alpha-roughness in the bitangent direction.
+//
+// This is from the `KHR_materials_anisotropy` spec:
+// <https://github.com/KhronosGroup/glTF/blob/main/extensions/2.0/Khronos/KHR_materials_anisotropy/README.md#image-based-lighting>
```
Comment (IceSentry) [inline crates/bevy_pbr/src/render/pbr_lighting.wgsl:163]: I think this is linking to the wrong section? At least the D_GGX_anistropic function is defined in this section:

```suggestion
// <https://github.com/KhronosGroup/glTF/blob/main/extensions/2.0/Khronos/KHR_materials_anisotropy/README.md#individual-lights>
```

File: crates/bevy_pbr/src/render/pbr_lighting.wgsl [hunk 6]
```diff
@@ -176,6 +229,27 @@ fn fresnel(f0: vec3<f32>, LdotH: f32) -> vec3<f32> {
...
+    V: f32,
+    F: vec3<f32>,
+    specular_intensity: f32,
+) -> vec3<f32> {
+    // Unpack.
+    let F0 = (*input).F0_;
+    let F_ab = (*input).F_ab;
+
+    var Fr = (specular_intensity * D * V) * F;
+    Fr *= 1.0 + F0 * (1.0 / F_ab.x - 1.0);
```
Comment (IceSentry) [inline crates/bevy_pbr/src/render/pbr_lighting.wgsl:249]: The linked Filament doc uses `dfg.y` instead `F_ab.x`. I assume `F_ab` is the same as `dfg`, but is the .x vs .y correct?

Comment (IceSentry) [reply crates/bevy_pbr/src/render/pbr_lighting.wgsl:249]: Ah, I see you just copied that from somewhere else. I guess we shouldn't touch this in this PR then.

File: crates/bevy_pbr/src/render/pbr_lighting.wgsl [hunk 7]
```diff
@@ -453,12 +570,16 @@ fn directional_light(light_id: u32, input: ptr<function, LightingInput>) -> vec3
...
+    let L = (*light).direction_to_light.xyz;
+    var derived_input = derive_lighting_input(N, V, L);
 
     let diffuse = diffuse_color * Fd_Burley(input, &derived_input);
 
+#ifdef STANDARD_MATERIAL_ANISOTROPY
+    let specular_light = specular_anisotropy(input, &derived_input, L, 1.0);
+#else   // STANDARD_MATERIAL_ANISOTROPY
     let specular_light = specular(input, &derived_input, 1.0);
+#endif
```
Comment (IceSentry) [inline crates/bevy_pbr/src/render/pbr_lighting.wgsl:582]:
```suggestion
#endif // STANDARD_MATERIAL_ANISOTROPY
```

File: examples/3d/anisotropy.rs [hunk 8]
```diff
@@ -0,0 +1,315 @@
...
+        // Choose the appropriate help text for the anisotropy toggle.
+        let material_variant_help_text = if self.anisotropy_enabled {
+            "Press Enter to disable anisotropy"
+        } else {
+            "Press Enter to enable anisotropy"
+        };
+
+        // Choose the appropriate help text for the light toggle.
+        let light_help_text = match self.light_mode {
+            LightMode::Directional => "Press Space to switch to an point light",
```
Comment (IceSentry) [inline examples/3d/anisotropy.rs:291]:
```suggestion
            LightMode::Directional => "Press Space to switch to a point light",
```

**Comment by pcwalton** [comment]
@IceSentry `volumetric_fog` runs fine for me now that I've updated this PR to `main`.

File: crates/bevy_pbr/src/render/pbr_lighting.wgsl (hunk 6 above)
Comment (pcwalton) [reply crates/bevy_pbr/src/render/pbr_lighting.wgsl:249]: Yeah, we might want to check this later but I just copy and pasted this one.

**Review by IceSentry — APPROVED** [review]
I ran a bunch of examples again and found no new regressions

LGTM

**Comment by alice-i-cecile** [comment]
Thank you to everyone involved with the authoring or reviewing of this PR! This work is relatively important and needs release notes! Head over to https://github.com/bevyengine/bevy-website/issues/1346 if you'd like to help out.
